share/shwap/namespaced_data.go

```diff
@@ -0,0 +1,193 @@
+package shwap
+
+import (
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/celestiaorg/celestia-app/pkg/wrapper"
+	"github.com/celestiaorg/nmt"
+	nmt_pb "github.com/celestiaorg/nmt/pb"
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/share"
+	types_pb "github.com/celestiaorg/celestia-node/share/shwap/proto"
+)
+
+// NamespacedData stores collections of RowNamespaceData, each representing shares and their proofs
+// within a namespace.
+type NamespacedData []RowNamespaceData
+
+// Flatten combines all shares from all rows within the namespace into a single slice.
+func (ns NamespacedData) Flatten() []share.Share {
+	var shares []share.Share
+	for _, row := range ns {
+		shares = append(shares, row.Shares...)
+	}
+	return shares
+}
+
+// RowNamespaceData holds shares and their corresponding proof for a single row within a namespace.
+type RowNamespaceData struct {
+	Shares []share.Share `json:"shares"` // Shares within the namespace.
+	Proof  *nmt.Proof    `json:"proof"`  // Proof of the shares' inclusion in the namespace.
+}
```

Comment:
Row requires a constructor because it uses private fields. However, this type can be constructed using struct literals. I generally prefer to avoid adding constructor here, as it ensures a single method of creating the type rather than having two. This approach should improve consistency in how the type is created across the codebase.

Reply:
>  I generally prefer to avoid adding constructor here, as it ensures a single method of creating the type rather than having two.

Having a constructor does not prohibit one from instantiating the type using struct literal. It does not ensure a single method of creating. Constructors for such type is the pattern we use around the codebase, and I want us to be consistent with it

Reply:
I mean that if we add a constructor, it will allow two ways to do the same thing—constructing an object. I would prefer to limit it to a single method (by not adding the constructor) to ensure consistency.

>Constructors for such types are the pattern we use around the codebase, and I want us to be consistent with it.

However, the use of such a constructor is not a pattern in our codebase, which suggests it would be better to stick with a single constructor option.

Reply:
Also we do not have constructors for similar types in our code base. For example for similar pre-shwap types in existing code base:
- NamespacedRow doesn't have it https://github.com/celestiaorg/celestia-node/blob/0c019873b7ee4822ac86a007b5b68090656c7ce4/share/getter.go#L54-L58
- ShareWithProof doesn't have it too https://github.com/celestiaorg/celestia-node/blob/0c019873b7ee4822ac86a007b5b68090656c7ce4/share/eds/byzantine/share_proof.go#L24-L31

Reply:
These two are exceptions and not the rule. Especially the NamespacedRow. Lets add constructors

Reply:
To ensure we adopt the most effective pattern for our codebase, I'd like to explore this topic further with some specific questions:

- From reviewing our codebase, it appears that types similar to what we're implementing, such as `NamespacedRow` and `ShareWithProof`, do not use constructors and have been effective in their functionality. Could you provide examples where constructors are the standard practice for similar types (all fields exported + no logic in constructor)? This would help identify a clear pattern for us to follow.
- I'm interested in understanding the benefits you see in introducing constructors for these types. Why do you believe a constructor would be better for our current implementation? Knowing the specific advantages could clarify why this approach is necessary.